### PR TITLE
Refactor CI and Mirror to PNNL

### LINKED ISSUE #2: Set up CI pipelines for HyKKT
As HyKKT is moving to GitHub as its long term home, it would be good to configure CI pipelines there. See #1.

### PR DESCRIPTION
**Pull request by ryandanehy**
This refactors CI to use extends instead of anchors and addresses some [concerns](https://gitlab.pnnl.gov/exasgd/solvers/hykkt/-/issues/49) such as the inclusion of `needs[]` to some of the jobs. This also sets up [CI pipelines](https://gitlab.pnnl.gov/exasgd/solvers/hykkt/-/issues/52) and allows us to migrate the code to github. It will work by having github use actions to mirror the repo over at PNNL gitlab. Everytime a push is done to the github repo it than triggers a pipeline for testing on pnnl hpc reasources. Next steps will be actual migration to github + creation of github secrets such as GIT_USERNAME, GIT_PASSWORD, AND GIT_REPO_URL @CameronRutherford  I would love to hear your thoughts on this draft before proceeding. 

Fixes #2.

### CODE REVIEW

File: .github/workflow/push_mirror.yaml
Comment (cameronrutherford) [inline .github/workflow/push_mirror.yaml]:
```suggestion
on: [push, delete]
```

File: .github/workflow/push_mirror.yaml
Comment (cameronrutherford) [inline .github/workflow/push_mirror.yaml]: There should be an action after this that triggers the CI pipeline since you explicitly skip ci here.

File: .gitlab-ci.yml
Comment (cameronrutherford) [inline .gitlab-ci.yml]:
```suggestion
```

File: .gitlab-ci.yml
Comment (cameronrutherford) [inline .gitlab-ci.yml:106]: Missing `on_failure` clause for this job.

File: .gitlab-ci.yml
Comment (cameronrutherford) [inline .gitlab-ci.yml]: This is undefined elsewhere

File: .gitlab-ci.yml
Comment (cameronrutherford) [inline .gitlab-ci.yml]: `STATUS_PROEJCT` needs to be defined for this job template.

File: .gitlab-ci.yml
Comment (cameronrutherford) [inline .gitlab-ci.yml]: These tags should be added back

File: .gitlab-ci.yml
Comment (cameronrutherford) [inline .gitlab-ci.yml]: None of these rules do anything, and can be deleted.

File: .gitlab-ci.yml
Comment (cameronrutherford) [inline .gitlab-ci.yml:57]: Let's change this to be a dynamic set of partitions in the future.

File: .gitlab-ci.yml
Comment (cameronrutherford) [inline .gitlab-ci.yml]: I don't think we need these for the HyKKT build system

**Comment by ryandanehy** [comment]
@CameronRutherford when you get a chance can it get a second look?

File: .gitlab-ci.yml
Comment (cameronrutherford) [inline .gitlab-ci.yml]: I think this needs to be the full path in the GitLab group to the project: `exasgd/solvers/hykkt` or something of the sort

File: .gitlab-ci.yml
Comment (cameronrutherford) [inline .gitlab-ci.yml:61]: This job doesn't also extend the tags template?

FWIW you can push to the PNNL GitLab directly still to test out pipelines

File: .github/workflow/push_mirror.yaml
Comment (cameronrutherford) [inline .github/workflow/push_mirror.yaml]:
```suggestion
```
None of this is necessary as it doesn't change the result of the pipeline, and we don't need the pipeline_id in this context

**Review by cameronrutherford — CHANGES_REQUESTED** [review]
Just a few more changes

File: .gitlab-ci.yml
Comment (cameronrutherford) [reply .gitlab-ci.yml:57]: We could even have a list of SLURM partitions for now as discussed offline

Comment (cameronrutherford) [reply .gitlab-ci.yml:57]: Also maybe we want to build on the newer Newell box (newell02)

File: .gitlab-ci.yml
Comment (ryandanehy) [reply .gitlab-ci.yml]: maybe I am missing something but those tags have already been added back in the most recent mr fix commit. I think you were looking at the wrong commit

File: .gitlab-ci.yml
Comment (ryandanehy) [reply .gitlab-ci.yml]: maybe I am missing something but those rules have already been remove in the most recent mr fix commit. I think you were looking at the wrong commit.

File: .gitlab-ci.yml
Comment (ryandanehy) [reply .gitlab-ci.yml]: Are you referring to output_Suffix and script_args because they were deleted in the most recent commit.

File: .gitlab-ci.yml
Comment (cameronrutherford) [inline .gitlab-ci.yml]: Did you delete the test script? Make sure there is actually a real test being run here before merging, but having a smoke test of a pipeline might be good for debugging secrets...

Comment (ryandanehy) [reply .gitlab-ci.yml]: Good Idea. I will make a little smoke test for debugging secrets in the event something goes wrong.

File: .gitlab-ci.yml
Comment (ryandanehy) [reply .gitlab-ci.yml:57]: I am unsure how to build on the newer newell box. What partition do I specify?

Comment (pelesh) [reply .gitlab-ci.yml:57]: Let's not spend too much time on this now. All we need right now is a single working CI pipeline. Once we have everything working, we can consider "upgrades" 😄.

Comment (cameronrutherford) [reply .gitlab-ci.yml:57]: Agreed

File: .gitlab-ci.yml
Comment (cameronrutherford) [reply .gitlab-ci.yml]: This seems to be the last unresolved thread before we can experiment with getting secrets to work

Comment (ryandanehy) [reply .gitlab-ci.yml]: I am currently testing on the gitlab side first.

**Comment by ryandanehy** [comment]
@CameronRutherford and @pelesh I believe this is good to go. We might have to enable the workflow manually through the settings to get it to run. @pelesh can you check under the actions settings if the workflow called push_mirror is enabled.

**Comment by cameronrutherford** [comment]
> @CameronRutherford and @pelesh I believe this is good to go. We might have to enable the workflow manually through the settings to get it to run. @pelesh can you check under the actions settings if the workflow called push_mirror is enabled. 

I suggest just hopping on a quick call to get this working, but LGTM!

**Comment by ryandanehy** [comment]
Pipelines appear to be up and running now. See gitlab for proof as well https://gitlab.pnnl.gov/exasgd/solvers/hykkt/-/pipelines/20164/builds. I think this is ready or at least very close to being ready to be merged.

File: .gitlab-ci.yml
Comment (cameronrutherford) [inline .gitlab-ci.yml]: Lets remove this TODO since things seem to be working

File: .gitlab-ci.yml
Comment (cameronrutherford) [inline .gitlab-ci.yml]: Let's add this cleanup stage back in and switch to using the extends syntax.

File: .gitlab-ci.yml
Comment (cameronrutherford) [inline .gitlab-ci.yml]: Seems like linting is no longer being applied. Do we plan on using pre-commit hooks to enforce this on GitHub? If so I would recommend creating an issue to keep track of that.

File: .gitlab-ci.yml
Comment (cameronrutherford) [inline .gitlab-ci.yml]: Workflow reules are unnecessary here since you trigger the pipelines manually, and push with the skip-ci option in the push mirror

Comment (cameronrutherford) [reply .gitlab-ci.yml]: More to the point, GitLab has no knowledge of the state of GitHub merge requests, so the rules are meaningless in this case.

File: .gitlab-ci.yml
Comment (cameronrutherford) [inline .gitlab-ci.yml]: Pedantic: I don't think this comment needed to be moved.

File: .github/workflows/push_mirror.yaml
Comment (cameronrutherford) [inline .github/workflows/push_mirror.yaml:3]: This comment is no longer accurate

File: .gitlab-ci.yml
Comment (cameronrutherford) [inline .gitlab-ci.yml:43]: Pedantic: iirc there is a deception tag as well as a marianas tag.

File: .clang-format
Comment (cameronrutherford) [inline .clang-format:1]: While the cleanup of this file is welcome, we don't run linting anywhere anymore :)

**Review by cameronrutherford — CHANGES_REQUESTED** [review]
Small changes requested. Make sure to have a successful pipeline run on your final commit before merging.

**Comment by cameronrutherford** [comment]
I didn't notice this when I submitted the review, but I don't see any PNNL pipeline status being posted to the GitHub PR, so perhaps something still isn't working correctly.

File: .clang-format
Comment (pelesh) [reply .clang-format:1]: Let's add linting in a separate PR.

File: .gitlab-ci.yml
Comment (pelesh) [reply .gitlab-ci.yml]: This is out of scope for this PR. A separate issue will be submitted.

File: .gitlab-ci.yml
Comment (pelesh) [reply .gitlab-ci.yml]: I think this needs to be addressed before this PR is approved.

**Review by pelesh — CHANGES_REQUESTED** [review]
Just a couple of small things are needed per @CameronRutherford's suggestions before this PR is merged.

The CI pipeline status needs to be posted back on GitHub.

**Review by pelesh — APPROVED** [review]
Looks good!

**Pull request merged**
